File: netDxf/IO/DxfWriter.SectionManager.cs
// Copyright (c) netDxf contributors. Licensed under the MIT License.
using System.IO;
using System.Linq;
using netDxf.Collections;
using netDxf.Objects;

namespace netDxf.IO
{
    internal sealed partial class DxfWriter
    {
        private bool WriteSectionManagerPayload(DxfDatabaseObject item)
        {
            if (!(item is DxfStoredSectionManager manager)) return false;
            foreach (DxfTag tag in manager.Tags) this.WriteDatabaseTag(tag, false);
            return true;
        }

        private void PrepareSectionManagerClasses(DxfClassCollection definitions)
        {
            foreach (string name in new[] { "SECTION_MANAGER", "SECTIONMANAGER" })
            {
                bool typed = this.doc.Objects.Items.Any(item => item is DxfStoredSectionManager && item.CodeName == name);
                if (!typed) continue;
                int count = this.doc.Objects.Items.Count(item => item.CodeName == name);
                if (definitions.Contains(name))
                {
                    DxfClass definition = definitions[name];
                    if (definition.CppClassName != "AcDbSectionManager" || definition.IsEntity)
                        throw new InvalidDataException("CLASS conflicts with a stored section manager.");
                    definition.InstanceCount = count;
                }
                else definitions.Add(new DxfClass(name, "AcDbSectionManager", "ObjectDBX Classes")
                    { ProxyFlags = 1024, IsEntity = false, InstanceCount = count });
            }
        }
    }
}

File: javascript/netDxf/IO/DxfWriter.SectionManager.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import { DxfStoredSectionManager } from '../Objects/DxfStoredSectionManager.js';
import { DxfClass } from '../DxfClass.js';
import { InvalidDataException } from '../../runtime/Errors.js';
export function WriteSectionManagerPayload(chunk,version,item) {
  if(!(item instanceof DxfStoredSectionManager))return false;
  for(const tag of item.Tags)chunk.Write(tag.Code,tag.Value);return true;
}
export function PrepareSectionManagerClasses(document,definitions) {
  for(const name of ['SECTION_MANAGER','SECTIONMANAGER']) {
    if(!Array.from(document.Objects.Items).some(item=>item instanceof DxfStoredSectionManager && item.CodeName===name))continue;
    const count=Array.from(document.Objects.Items).filter(item=>item.CodeName===name).length;
    if(definitions.Contains(name)) {
      const definition=definitions.get_Item(name);
      if(definition.CppClassName!=='AcDbSectionManager' || definition.IsEntity)throw new InvalidDataException('CLASS conflicts with a stored section manager.');
      definition.InstanceCount=count;
    }else{
      const definition=new DxfClass(name,'AcDbSectionManager','ObjectDBX Classes');definition.ProxyFlags=1024;definition.IsEntity=false;definition.InstanceCount=count;definitions.Add(definition);
    }
  }
}
